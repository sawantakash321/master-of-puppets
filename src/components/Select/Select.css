.select {
  display: inline-block;
  position: relative; }
  .select select {
    display: inline-block;
    padding: 18px 20px 18px 15px;
    border: solid 1px #333366;
    border-radius: 3px;
    width: 100%;
    background-color: #ffffff; }
    .select select option.placeholder {
      color: lightgrey; }
  .select::before {
    position: absolute;
    content: ' ';
    top: 0;
    right: 0;
    bottom: 0;
    width: 40px;
    border: solid 1px #333366;
    border-left: none;
    background-color: #ffffff;
    border-top-right-radius: 3px;
    border-bottom-right-radius: 3px;
    pointer-events: none; }
  .select::after {
    position: absolute;
    right: 6px;
    top: 50%;
    content: ' ';
    width: 0;
    height: 0;
    border: solid 5px transparent;
    border-top: solid 10px #333366;
    border-bottom: solid 10px transparent;
    transform: translateY(-4px);
    pointer-events: none; }
  .select.open::after {
    transform: rotateX(180deg) translateY(16px); }
