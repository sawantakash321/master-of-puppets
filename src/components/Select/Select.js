import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

class Select extends Component {
  constructor(props) {
    super(props);

    this.state = { isOpen: false };
    this.onToggle = this.onToggle.bind(this);
    this.onBlur = this.onBlur.bind(this);
  }

  onToggle() {
    this.setState(({ isOpen, ...otherState }) => {
      return {
        ...otherState,
        isOpen: !isOpen
      };
    });
  }

  onBlur() {
    this.setState(prevState => {
      return {
        ...prevState,
        isOpen: false
      };
    });
  }

  render() {
    const {
      placeholder = '',
      items = [],
      onChange,
      value,
      id,
      name
    } = this.props;
    const { isOpen } = this.state;

    // Compose the items by converting strings or objects into option elements
    const options = items.map(i => {
      if (typeof i === 'string') {
        return (
          <option key={i} value={i}>
            {i}
          </option>
        );
      }

      return (
        <option key={i.value} value={i.value}>
          {i.name}
        </option>
      );
    });

    return (
      <div className={classNames({ select: true, open: isOpen })}>
        <select
          onClick={this.onToggle}
          value={value}
          id={id}
          name={name}
          onBlur={this.onBlur}
          onChange={onChange}
        >
          {placeholder && (
            <option className="placeholder" value="">
              {placeholder}
            </option>
          )}
          {options}
        </select>
      </div>
    );
  }
}

Select.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  placeholder: PropTypes.string,
  items: PropTypes.array,
  value: PropTypes.any,
  onChange: PropTypes.func
};

export default Select;
