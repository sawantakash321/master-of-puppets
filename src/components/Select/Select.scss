$select-bg-color: #ffffff;

.select {
  display: inline-block;

  position: relative;

  select {
    display: inline-block;
    padding: 18px 20px 18px 15px;
    border: solid 1px #333366;
    border-radius: 3px;
    width: 100%;
    background-color: $select-bg-color;

    // This is the empty option element that serves as placeholder
    option.placeholder {
      color: lightgrey;
    }
  }

  &::before {
    position: absolute;
    content: ' ';
    top: 0;
    right: 0;
    bottom: 0;
    width: 40px;
    border: solid 1px #333366;
    border-left: none;
    background-color: $select-bg-color;
    border-top-right-radius: 3px;
    border-bottom-right-radius: 3px;
    pointer-events: none;
  }

  &::after {
    position: absolute;
    right: 6px;
    top: 50%;
    content: ' ';
    width: 0;
    height: 0;
    border: solid 5px transparent;
    border-top: solid 10px #333366;
    border-bottom: solid 10px transparent;
    transform: translateY(-4px);
    pointer-events: none;
  }

  &.open::after {
    transform: rotateX(180deg) translateY(16px);
  }
}
