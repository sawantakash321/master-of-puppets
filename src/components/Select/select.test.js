/* global jest, describe, expect, it */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import Select from './Select';

describe('Select component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Select />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders a placeholder', () => {
    const select = shallow(<Select placeholder="test" />);
    const placeholder = select.find('option.placeholder');
    expect(placeholder.length).toEqual(1);
    expect(placeholder.hasClass('placeholder')).toBeTruthy();
    expect(placeholder.html()).toEqual(
      '<option class="placeholder" value="">test</option>'
    );
  });

  it('renders one option for each element from the items array', () => {
    const items = ['item1', 'item2'];
    const select = shallow(<Select items={items} />);
    const options = select.find('option');
    expect(options.length).toEqual(2);
    options.forEach((o, idx) => {
      expect(o.text()).toEqual(items[idx]);
    });
  });

  it('renders options when the items are objects and not strings', () => {
    const items = [{ name: 'item1', value: 1 }, { name: 'item2', value: 2 }];
    const select = shallow(<Select items={items} />);
    const options = select.find('option');
    expect(options.length).toEqual(2);
    options.forEach((o, idx) => {
      const item = items[idx];
      expect(o.html()).toEqual(
        `<option value="${item.value}">${item.name}</option>`
      );
    });
  });

  it('renders options and the placeholder', () => {
    const items = ['item1', 'item2'];
    const select = shallow(<Select items={items} placeholder="test" />);
    const options = select.find('option');
    expect(options.length).toEqual(3);
    options.forEach((o, idx) => {
      if (idx === 0) {
        expect(o.html()).toEqual(
          `<option class="placeholder" value="">test</option>`
        );
      } else {
        const item = items[idx - 1];
        expect(o.html()).toEqual(`<option value="${item}">${item}</option>`);
      }
    });
  });

  it('renders sets form related attributes', () => {
    const select = mount(<Select id="select1" name="select2" />);
    const htmlSelect = select.find('select');
    expect(htmlSelect.length).toEqual(1);
    const node = htmlSelect.getDOMNode();
    expect(node.attributes.getNamedItem('id').value).toEqual('select1');
    expect(node.attributes.getNamedItem('name').value).toEqual('select2');
  });

  it('toggles the open class when the select is clicked', () => {
    const select = shallow(<Select items={['item1', 'item2']} />);
    const htmlSelect = select.find('select');
    htmlSelect.simulate('click');
    expect(select.hasClass('open')).toBeTruthy();
    htmlSelect.simulate('click');
    expect(select.hasClass('open')).toBeFalsy();
  });

  it('removes the open class when the select loses focus', () => {
    const select = shallow(<Select items={['item1', 'item2']} />);
    const htmlSelect = select.find('select');
    htmlSelect.simulate('click');
    expect(select.hasClass('open')).toBeTruthy();
    htmlSelect.simulate('blur');
    expect(select.hasClass('open')).toBeFalsy();
  });

  it('removes the open class when the select loses focus', () => {
    const select = shallow(<Select items={['item1', 'item2']} />);
    const htmlSelect = select.find('select');
    htmlSelect.simulate('click');
    expect(select.hasClass('open')).toBeTruthy();
    htmlSelect.simulate('blur');
    expect(select.hasClass('open')).toBeFalsy();
  });

  // Snapshot test
  it('matches the snapshot', () => {
    const props = {
      placeholder: 'placeholder',
      id: 'select',
      name: 'select',
      value: 'first option',
      items: ['first option', 'second option']
    };
    const events = {
      onChange: jest.fn()
    };
    const tree = renderer.create(<Select {...props} {...events} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
