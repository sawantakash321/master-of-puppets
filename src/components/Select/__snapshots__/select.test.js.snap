// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component matches the snapshot 1`] = `
<div
  className="select"
>
  <select
    id="select"
    name="select"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    value="first option"
  >
    <option
      className="placeholder"
      value=""
    >
      placeholder
    </option>
    <option
      value="first option"
    >
      first option
    </option>
    <option
      value="second option"
    >
      second option
    </option>
  </select>
</div>
`;
